src/space_tracker.rs

```diff
@@ -0,0 +1,80 @@
+// Copyright (c) 2024-present, fjall-rs
+// This source code is licensed under both the Apache 2.0 and MIT License
+// (found in the LICENSE-* files in the repository)
+
+use std::sync::atomic::{AtomicU64, Ordering};
+
+/// Keeps track of disk space or buffer size in bytes
+#[derive(Debug)]
+pub struct SpaceTracker(AtomicU64);
+
+impl SpaceTracker {
+    pub fn new() -> SpaceTracker {
+        Self(AtomicU64::new(0))
+    }
+
+    pub fn get(&self) -> u64 {
+        self.0.load(Ordering::Acquire)
+    }
+
+    // Adds some bytes to the space counter.
+    //
+    // Returns the counter *after* incrementing.
+    pub fn allocate(&self, n: u64) -> u64 {
+        use Ordering::{Acquire, SeqCst};
+
+        loop {
+            let now = self.0.load(Acquire);
+            let added = now.saturating_add(n);
+
+            if self.0.compare_exchange(now, added, SeqCst, SeqCst).is_ok() {
+                return added;
+            }
+        }
```

Comment:
For `free()` I originally used CAS for `saturating_sub` semantics in case there is a logic error, because a negative wraparound would be disastrous. However, for the `allocate()` case, we are never gonna reach 18 trillion whatever bytes anyway, so `fetch_add` is fine I think.